File: backend_node/api/areas.js
const express = require('express');
const database = require('../models/database');
const { getProjectFiles } = require('../utils/fileUtils');

const router = express.Router();

router.get('/areas', async (req, res) => {
    try {
        const page = parseInt(req.query.page) || 1;
        const perPage = parseInt(req.query.per_page) || 10;
        
        // Filters
        const filters = {
            idFilter: req.query.id_filter || '',
            projectIdFilter: req.query.project_id_filter || '',
            xminFilter: req.query.xmin_filter || '',
            yminFilter: req.query.ymin_filter || '',
            xmaxFilter: req.query.xmax_filter || '',
            ymaxFilter: req.query.ymax_filter || '',
            scaleFilter: req.query.scale_filter || ''
        };

        let whereConditions = [];
        let params = [];

        // Build WHERE conditions
        if (filters.idFilter) {
            const idVal = parseInt(filters.idFilter);
            if (!isNaN(idVal)) {
                whereConditions.push('a.id = ?');
                params.push(idVal);
            } else {
                whereConditions.push('a.id = ?');
                params.push(-1); // Invalid condition to return no results
            }
        }
        if (filters.projectIdFilter) {
            whereConditions.push('a.project_id LIKE ?');
            params.push(`%${filters.projectIdFilter}%`);
        }
        if (filters.xminFilter) {
            const xminVal = parseFloat(filters.xminFilter);
            if (!isNaN(xminVal)) {
                whereConditions.push('a.xmin = ?');
                params.push(xminVal);
            } else {
                whereConditions.push('a.xmin = ?');
                params.push(-1); // Invalid condition
            }
        }
        if (filters.yminFilter) {
            const yminVal = parseFloat(filters.yminFilter);
            if (!isNaN(yminVal)) {
                whereConditions.push('a.ymin = ?');
                params.push(yminVal);
            } else {
                whereConditions.push('a.ymin = ?');
                params.push(-1); // Invalid condition
            }
        }
        if (filters.xmaxFilter) {
            const xmaxVal = parseFloat(filters.xmaxFilter);
            if (!isNaN(xmaxVal)) {
                whereConditions.push('a.xmax = ?');
                params.push(xmaxVal);
            } else {
                whereConditions.push('a.xmax = ?');
                params.push(-1); // Invalid condition
            }
        }
        if (filters.ymaxFilter) {
            const ymaxVal = parseFloat(filters.ymaxFilter);
            if (!isNaN(ymaxVal)) {
                whereConditions.push('a.ymax = ?');
                params.push(ymaxVal);
            } else {
                whereConditions.push('a.ymax = ?');
                params.push(-1); // Invalid condition
            }
        }
        if (filters.scaleFilter) {
            whereConditions.push('a.scale LIKE ?');
            params.push(`%${filters.scaleFilter}%`);
        }

        const whereClause = whereConditions.length > 0 ? `WHERE ${whereConditions.join(' AND ')}` : '';
        
        // Get total count for pagination
        const countQuery = `SELECT COUNT(*) as total FROM areas a ${whereClause}`;
        const countResult = await database.get(countQuery, params);
        const totalItems = countResult.total;
        const totalPages = Math.ceil(totalItems / perPage);

        if (page > totalPages && totalPages > 0) {
            return res.status(400).json({
                error: 'Page number exceeds total pages',
                total_pages: totalPages,
                current_page: page
            });
        }

        // Main query with pagination
        const offset = (page - 1) * perPage;
        const query = `
            SELECT 
                a.*,
                p.project_name,
                p.file_location
            FROM areas a
            LEFT JOIN projects p ON a.project_id = p.uuid
            ${whereClause}
            ORDER BY a.id
            LIMIT ? OFFSET ?
        `;
        const queryParams = [...params, perPage, offset];

        const areas = await database.all(query, queryParams);

        // Enhance areas with file information
        const enhancedAreas = [];
        for (const area of areas) {
            try {
                if (area.file_location) {
                    const fileInfo = getProjectFiles(area.file_location);
                    enhancedAreas.push({
                        ...area,
                        file_count: fileInfo.file_count,
                        most_recent_file: fileInfo.most_recent,
                        all_files: fileInfo.all_files,
                        // Add frontend-expected properties
                        project_all_files: fileInfo.all_files,
                        project_file_location: area.file_location
                    });
                } else {
                    enhancedAreas.push({
                        ...area,
                        file_count: 0,
                        most_recent_file: null,
                        all_files: [],
                        // Add frontend-expected properties
                        project_all_files: [],
                        project_file_location: area.file_location || ''
                    });
                }
            } catch (error) {
                console.warn(`Error getting file info for area ${area.id}:`, error.message);
                enhancedAreas.push({
                    ...area,
                    file_count: 0,
                    most_recent_file: null,
                    all_files: [],
                    // Add frontend-expected properties
                    project_all_files: [],
                    project_file_location: area.file_location || ''
                });
            }
        }

        res.json({
            areas: enhancedAreas,
            pagination: {
                page: page,
                per_page: perPage,
                total_items: totalItems,
                total_pages: totalPages,
                has_prev: page > 1,
                has_next: page < totalPages
            },
            filters: filters
        });

    } catch (error) {
        console.error('Error in get_all_areas:', error);
        res.status(500).json({ error: 'Internal server error' });
    }
});

router.get('/areas/:id', async (req, res) => {
    try {
        const id = parseInt(req.params.id);
        
        if (isNaN(id)) {
            return res.status(400).json({ error: 'Invalid area ID' });
        }

        const query = `
            SELECT 
                a.*,
                p.project_name,
                p.file_location
            FROM areas a
            LEFT JOIN projects p ON a.project_id = p.uuid
            WHERE a.id = ?
        `;

        const area = await database.get(query, [id]);
        
        if (!area) {
            return res.status(404).json({ error: 'Area not found' });
        }

        // Get file information
        try {
            if (area.file_location) {
                const fileInfo = getProjectFiles(area.file_location);
                area.file_count = fileInfo.file_count;
                area.most_recent_file = fileInfo.most_recent;
                area.all_files = fileInfo.all_files;
                // Add frontend-expected properties
                area.project_all_files = fileInfo.all_files;
                area.project_file_location = area.file_location;
            } else {
                area.file_count = 0;
                area.most_recent_file = null;
                area.all_files = [];
                // Add frontend-expected properties
                area.project_all_files = [];
                area.project_file_location = '';
            }
        } catch (error) {
            console.warn(`Error getting file info for area ${id}:`, error.message);
            area.file_count = 0;
            area.most_recent_file = null;
            area.all_files = [];
            // Add frontend-expected properties
            area.project_all_files = [];
            area.project_file_location = area.file_location || '';
        }

        res.json(area);

    } catch (error) {
        console.error('Error in get_area:', error);
        res.status(500).json({ error: 'Internal server error' });
    }
});

module.exports = router;

File: backend/api/areas.py
from flask import Blueprint, jsonify, request
from sqlalchemy import select, func, and_
from models.database import engine, areas_table, projects_table
from utils.file_utils import get_project_files
import os

areas_bp = Blueprint('areas', __name__)

@areas_bp.route('/areas', methods=['GET'])
def get_all_areas():
    """Get all areas with pagination and filtering"""
    try:
        page = request.args.get('page', 1, type=int)
        per_page = request.args.get('per_page', 10, type=int)
        
        # Filters
        filters = {}
        filters['id_filter'] = request.args.get('id_filter', '', type=str)
        filters['project_id_filter'] = request.args.get('project_id_filter', '', type=str)
        filters['xmin_filter'] = request.args.get('xmin_filter', '', type=str)
        filters['ymin_filter'] = request.args.get('ymin_filter', '', type=str)
        filters['xmax_filter'] = request.args.get('xmax_filter', '', type=str)
        filters['ymax_filter'] = request.args.get('ymax_filter', '', type=str)
        filters['scale_filter'] = request.args.get('scale_filter', '', type=str)
        
        query_filters = []
        if filters['id_filter']:
            try:
                id_val = int(filters['id_filter'])
                query_filters.append(areas_table.c.id == id_val)
            except ValueError:
                query_filters.append(areas_table.c.id == -1)
        if filters['project_id_filter']:
            query_filters.append(areas_table.c.project_id.ilike(f"%{filters['project_id_filter']}%"))
        if filters['xmin_filter']:
            try:
                xmin_val = float(filters['xmin_filter'])
                query_filters.append(areas_table.c.xmin == xmin_val)
            except ValueError:
                query_filters.append(areas_table.c.xmin == -1)
        if filters['ymin_filter']:
            try:
                ymin_val = float(filters['ymin_filter'])
                query_filters.append(areas_table.c.ymin == ymin_val)
            except ValueError:
                query_filters.append(areas_table.c.ymin == -1)
        if filters['xmax_filter']:
            try:
                xmax_val = float(filters['xmax_filter'])
                query_filters.append(areas_table.c.xmax == xmax_val)
            except ValueError:
                query_filters.append(areas_table.c.xmax == -1)
        if filters['ymax_filter']:
            try:
                ymax_val = float(filters['ymax_filter'])
                query_filters.append(areas_table.c.ymax == ymax_val)
            except ValueError:
                query_filters.append(areas_table.c.ymax == -1)
        if filters['scale_filter']:
            try:
                # Try to parse as float for backward compatibility
                scale_val = float(filters['scale_filter'])
                query_filters.append(areas_table.c.scale == str(scale_val))
            except ValueError:
                # If not a number, treat as string scale format
                query_filters.append(areas_table.c.scale.ilike(f"%{filters['scale_filter']}%"))

        with engine.connect() as conn:
            # Get total count for areas pagination
            count_stmt = select(func.count()).select_from(areas_table)
            if query_filters:
                count_stmt = count_stmt.where(and_(*query_filters))
            total_items = conn.execute(count_stmt).scalar_one()

            total_pages = (total_items + per_page - 1) // per_page
            if page > total_pages and total_pages > 0:
                page = total_pages
            elif total_pages == 0:
                page = 1

            # Query areas for the current page with filters, joined with projects to get file location
            stmt = select(
                areas_table.c.id, 
                areas_table.c.project_id, 
                areas_table.c.xmin, 
                areas_table.c.ymin, 
                areas_table.c.xmax, 
                areas_table.c.ymax, 
                areas_table.c.scale, 
                projects_table.c.file_location.label('project_file_location')
            )
            stmt = stmt.select_from(areas_table.join(projects_table, areas_table.c.project_id == projects_table.c.uuid))
            
            if query_filters:
                stmt = stmt.where(and_(*query_filters))
            
            stmt = stmt.limit(per_page).offset((page - 1) * per_page)
            areas = conn.execute(stmt).fetchall()

            # Add file information for areas (show files of associated project)
            areas_list = []
            for area in areas:
                area_dict = dict(area)
                project_file_location = area_dict['project_file_location']
                
                # Add file information
                file_info = get_project_files(project_file_location)
                area_dict['project_all_files'] = file_info['all_files']
                area_dict['project_file_count'] = file_info['file_count']
                
                # Add absolute file location
                abs_path = os.path.abspath(project_file_location)
                area_dict['project_abs_file_location'] = abs_path
                
                if file_info['most_recent']:
                    area_dict['project_view_file_path'] = file_info['most_recent']['rel_path']
                    area_dict['project_view_file_type'] = file_info['most_recent']['type']
                else:
                    area_dict['project_view_file_path'] = None
                    area_dict['project_view_file_type'] = None

                areas_list.append(area_dict)

            return jsonify({
                'areas': areas_list,
                'pagination': {
                    'current_page': page,
                    'per_page': per_page,
                    'total_pages': total_pages,
                    'total_items': total_items
                },
                'filters': filters
            })

    except Exception as e:
        return jsonify({'error': str(e)}), 500
